src/hotspot/share/opto/loopPredicate.cpp

```diff
@@ -1505,30 +1513,31 @@ IfProjNode* PhaseIdealLoop::add_template_assertion_predicate(IfNode* iff, IdealL
   // Second predicate for init + (current stride - initial stride)
   // This is identical to the previous predicate initially but as
   // unrolling proceeds current stride is updated.
-  Node* init_stride = loop->_head->as_CountedLoop()->stride();
-  Node* opaque_stride = new OpaqueLoopStrideNode(C, init_stride);
-  register_new_node(opaque_stride, new_proj);
-  Node* max_value = new SubINode(opaque_stride, init_stride);
-  register_new_node(max_value, new_proj);
-  max_value = new AddINode(opaque_init, max_value);
-  register_new_node(max_value, new_proj);
-  // init + (current stride - initial stride) is within the loop so narrow its type by leveraging the type of the iv Phi
-  max_value = new CastIINode(max_value, loop->_head->as_CountedLoop()->phi()->bottom_type());
-  register_new_node(max_value, predicate_proj);
+  if (max_value == nullptr) {
+    // init + (current stride - initial stride) is within the loop so narrow its type by leveraging the type of the iv Phi
+    iv_phi_assertion_predicate_condition(loop->_head->as_CountedLoop(), new_proj, opaque_init, max_value);
+    new_proj = add_template_assertion_predicate_helper(predicate_proj, reason, new_proj, bol, Op_If);
```

Comment:
Shouldn't this use the the `bol` returned by `iv_phi_assertion_predicate_condition`?

Reply:
Thank you for looking at this. And good catch! Some last minute refactoring gone wrong.